# READ BEFORE CONTINUE

- Untuk mendapatkan akses edit kalian bisa membuat copy di perangkat kalian masing-masing dengan menggunakan akun gmail masing-masing (Menu File -> Save copy in Drive)

# **Modul 1 Penggalian Data**

---
 DATA STRUCTURE IN PYTHON

Disusun oleh:
1. Alviya Laela Lestari
2. Kiara Azzahra

Diaudit oleh:
Yufis Azhar, S.Kom, M.Kom

Presented by: TIM LAB-IT Universitas Muhammadiyah Malang

## Tujuan Praktikum

1. Mahasiswa mampu memahami dasar bahasa pemrograman python
2. Mahasiswa mampu memahami tipe-tipe struktur data pada bahasa pemrograman python
3. Mahasiswa mampu memahami list, array, tuple dan dataframe


## Persiapan Software/Apilkasi

**1. Hardware**
- Laptop/Komputer

**2. Software (Pilih salah satu)**
- Visual Studio Code, Google Collaboratory, PyCharm, Jupyter Notebook

**3. Tutorial**
- Menggunakan Google Colab : https://youtu.be/tpDjhSzJor4?si=bFjSP_mf3oB5x520
- Menggunakan Jupyter di VS Code : https://youtu.be/HOg13z1wI1A?si=RhEY2QY5bir-dzEH  
- Menggunakan Jupyter Notebook : https://youtu.be/HQaCF2F_MRo?si=k6ubuGuCg-VS0tQe  
Instal Anaconda : https://youtu.be/XyudQ3B3OWI?si=7qugN_vTxAM30qDt


# 1. List dan Tuple

Pada semester sebelumnya, kalian sudah mempelajari terkait Pemrograman Fungsional yang membahas tipe data yang ada dalam Python, termasuk sequence type. Sequence adalah kumpulan nilai yang diakses melalui indeks dan memiliki beberapa tipe data utama, seperti array. Namun, di Python, tipe data sequence sedikit berbeda dari array pada umumnya. Data sequence lebih dinamis dibandingkan array, kita bisa menambahkan elemen sebanyak yang kita inginkan tanpa batasan ukuran seperti pada array. Selain itu, sequence juga dapat menyimpan elemen dengan tipe data yang berbeda.


Pada modul ini, kita akan membahas ulang dua tipe data sequence yang sangat sering digunakan, yaitu List dan Tuple. Keduanya memiliki fitur yang mirip namun dengan karakteristik tertentu yang membedakannya.

## **1.1 List**

List dalam Python adalah salah satu tipe data sequence yang paling fleksibel dan sering digunakan. Listi ini berisi kumpulan elemen yang berurutan (ter-index) dan dapat diubah (mutable). Berikut adalah beberapa poin penting tentang list dalam Python:

a. Mutable (Dapat Diubah)

Kita tetap bisa melakukan perubahan pada data dalam list karena sifat dari list yang dapat diubah/mutable.


b. Ordered (Berurutan)

Maksud dari berurutan disini adalah bahwa untuk mengakses data dalam list akan dimulai berurutan sesuai dengan urutan indexnya.

In [ ]:
list = ['a', 'b', 'c', 'd', 'e']
print(list[0])
print(list[1])
print(list[2])

a
b
c


c. Heterogeneous (Banyak Tipe Data)

List tidak kaku yang berarti List mampu menyimpan value dengan bermacam-macam tipe data seperti contoh di bawah ini:

In [ ]:
list = ["data", 7, 1.25, 8, "mining", 5, False]
print(list)

['data', 7, 1.25, 8, 'mining', 5, False]


d. Allow Duplicates (Duplikasi)

List memperbolehkan data yang memiliki value sama karena List dipengaruhi oleh index.

In [ ]:
list = [1, "adi", True, 1, "adi", True]
print(list)

[1, 'adi', True, 1, 'adi', True]


e. Dinamis

Oleh karena sifatnya yang muatable, list memperbolehkan operasi pada data. Operasi yang dapat digunakan pada list:

- Menambahkan item

extend() : menambahkan **semua elemen iterable** (seperti list, tuple atau string) **satu per satu** di posisi terakhir list.

In [ ]:
list = [1, "adi", True]
list.extend(["budi", 7, 2.15])
print(list)

[1, 'adi', True, 'budi', 7, 2.15]


append() : menambahkan **satu elemen yang diperlakukan sebagai satu objek tunggal** di posisi terakhir list.

In [ ]:
list = [1, "adi", True]
list.append(["budi", 7, 2.15])
print(list)

[1, 'adi', True, ['budi', 7, 2.15]]


insert() : menambahkan list baru ke posisi yang ditentukan.

Dalam kode di bawah, kita akan menyisipkan list baru sebelum indeks ke-2, yaitu sebelum elemen True.

In [ ]:
list = [1, "adi", True]
list.insert(2, ["budi", 7, 2.15])
print(list)

[1, 'adi', ['budi', 7, 2.15], True]


- Menghapus item

remove() : menghapus item pertama dalam list yang cocok dengan nilai yang diberikan.

In [ ]:
fruits = ["apple", "banana", "cherry", "apple"]
fruits.remove("apple")
print(fruits)

['banana', 'cherry', 'apple']


Jika perintah remove kembali dijalankan, maka item ’apple’ selanjutnya akan terhapus.

In [ ]:
fruits.remove("apple")
print(fruits)

['banana', 'cherry']


pop(index): menghapus item berdasarkan index dan mengembalikan item yang dihapus. Jika nilai index tidak diberikan, maka item terakhir yang akan dihapus.

In [ ]:
numbers = [10, 20, 30, 40, 50, 60, 70]

tanpa_index = numbers.pop()
print("Pop tanpa index:", numbers)
print("Removed:", tanpa_index)

index = numbers.pop(2)
print("Pop dengan index:", numbers)
print("Removed:", index)

Pop tanpa index: [10, 20, 30, 40, 50, 60]
Removed: 70
Pop dengan index: [10, 20, 40, 50, 60]
Removed: 30


clear(): menghapus seluruh item dari list, sehingga list menjadi kosong.

In [ ]:
names = ["Alice", "Bob", "Charlie"]
names.clear()
print(names)

[]


- Menemukan item

index(item): mencari index dari item pertama yang cocok dengan nilai yang diberikan.

In [ ]:
colors = ["red", "blue", "green", "blue"]
index = colors.index("blue")
print(index)

1


- Menghitung item

count(item): menghitung jumlah kemunculan item dalam list.

In [ ]:
scores = [10, 20, 10, 30, 10, 20, 10, 70, 10]
count = scores.count(10)
print(count)

# Output: 5

5


- Mengurutkan item

sort(): mengurutkan item dalam list secara ascending secara default, atau descending jika menggunakan parameter reverse=True.

In [ ]:
numbers = [10, 20, 10, 30, 10, 20, 10, 70, 10]
numbers.sort()  # Ascending
print(numbers)

numbers.sort(reverse=True)  # Descending
print(numbers)

[10, 10, 10, 10, 10, 20, 20, 30, 70]
[70, 30, 20, 20, 10, 10, 10, 10, 10]


- Membalik urutan

reverse(): membalikkan urutan elemen dalam list tanpa memperhatikan nilai elemen.

In [ ]:
letters = ["a", "b", "c", "d", 1, 4, True, 2, 3, 1]
letters.reverse()
print(letters)

[1, 3, 2, True, 4, 1, 'd', 'c', 'b', 'a']


## 1.2 Tuple

Tuple adalah urutan objek dalam Python yang bersifat **tidak dapat diubah (immutable).** Tuple serupa dengan list, namun perbedaan utamanya adalah elemen dalam tuple tidak dapat diubah setelah didefinisikan. Tuple menggunakan tanda kurung ‘()’, sedangkan list menggunakan tanda kurung siku ‘[]’. Membuat tuple sangat mudah; kita hanya perlu menuliskan nilai-nilai yang dipisahkan oleh koma. Berikut adalah beberapa poin penting tentang tuple dalam Python:

**a. Immutable(tidak dapat diubah)**

Saat berhasil dibuat, tuple tidak dapat diubah, termasuk menambahkan atau menghapus elemen di dalamnya.

In [ ]:
tuple = ("a", "b", "c", "d", "e")
tuple[0] = "z"
print(tuple)

TypeError: 'tuple' object does not support item assignment

**b. Heterogeneous (Banyak Tipe Data)**

Tuple mengizinkan untuk memiliki data yang berbeda tipe seperti List.


In [ ]:
tuple = ("data", 7, 1.25, 8, "mining", 5, False)
print(tuple)

('data', 7, 1.25, 8, 'mining', 5, False)


**c. Ordered (Berurutan)**

Tuple bersifat ordered seperti List yaitu untuk mengaksesnya kita bisa menggunakan index masing-masing data.

In [ ]:
tuple = (1, 2, 3, 4, 5)
print(tuple[1])
print(tuple[2])
print(tuple[3])

2
3
4


**d. Indexable**

Setiap item dalam tuple dapat diakses menggunakan indeksnya.

In [ ]:
t = (10, 20, 30, 40)
print(t[0])
print(t[3])

10
40


**e. Allow Duplicates (Duplikasi)**

Tuple juga memperbolehkan data yang memiliki value sama seperti List yang dipengaruhi oleh index.


In [ ]:
tuple = (True, 7, "budi", "budi", 7, True)
print(tuple)

(True, 7, 'budi', 'budi', 7, True)


# **2. NumPy dan Array**

NumPy (Numerical Python) adalah library Python yang digunakan untuk mengelola array berukuran besar, termasuk array multidimensi. Library ini juga menyediakan berbagai fungsi matematika untuk mempermudah perhitungan kompleks pada array. Array di NumPy adalah struktur data berbentuk grid yang dirancang khusus untuk menyimpan data numerik. Dibandingkan dengan list Python, array NumPy lebih cepat dan efisien dalam melakukan operasi perhitungan karena telah dioptimalkan.

## **2.1 Array NumPy vs List Python**

| **Array NumPy**                 | **List Python**                     |
|----------------------------------|--------------------------------------|
| Ukuran Tetap                    | Ukuran Dinamis                      |
| Elemen harus memiliki tipe data yang sama | Tipe elemen bisa beragam            |
| Operasi matematika element-wise lebih cepat | Tidak mendukung operasi langsung    |


## **2.2 Membuat Array di NumPy**

a. Array 1D

Membuat array satu dimensi yang digunakan untuk menyimpan daftar data linear.

In [ ]:
import numpy as np
arr = np.array([1, 2, 3, 4, 5])
print(arr)

[1 2 3 4 5]


b. Array 2D

Membuat array dua dimensi yang menyerupai matriks atau tabel data dengan baris dan kolom.

In [ ]:
matrix = np.array([[1, 2], [3, 4], [5, 6]])
print(matrix)

[[1 2]
 [3 4]
 [5 6]]


c. Array 3D

Membuat array tiga dimensi yang bisa menyimpan data kompleks seperti tensor dalam pembelajaran mesin.

In [ ]:
array_3d = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
print(array_3d)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


## **2.3 Properti Array**

**a. Dimensi Array (ndim)**

Memberikan informasi tentang jumlah dimensi array, misalnya 1D, 2D, atau 3D.

**b. Bentuk Array (shape)**

Menunjukkan jumlah elemen pada setiap dimensi (misalnya, array 2x3 memiliki shape (2, 3)).

**c. Jumlah Elemen (size)**

Mengembalikan total elemen dalam array.

**d. Tipe Data (dtype)**

Menjelaskan tipe data yang digunakan oleh array, seperti int32 atau float64.


In [ ]:
array = np.array([[1, 2, 3], [4, 5, 6]])

print("Dimensi:", array.ndim)
print("Shape:", array.shape)
print("Jumlah elemen:", array.size)
print("Tipe data:", array.dtype)

Dimensi: 2
Shape: (2, 3)
Jumlah elemen: 6
Tipe data: int64


## **2.4 Operasi Matematika pada Array**

**a. Operasi Element-wise**

Melakukan operasi matematika seperti penjumlahan, perkalian, atau pemangkatan pada elemen array secara langsung, elemen demi elemen.

In [ ]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

print(a + b)  # Output: [5 7 9]
print(a * b)  # Output: [4 10 18]
print(a ** 2) # Output: [1 4 9]

[3 4 5]
[ 4 10 18]
[1 4 9]


**b. Operasi pada Matriks**

Melakukan operasi aljabar linear  yang berguna untuk manipulasi data numerik skala besar.

- Penjumlahan dan Pengurangan pada Matriks

Penjumlahan dan pengurangan dua matriks hanya bisa dilakukan jika keduanya memiliki ukuran yang sama.

In [ ]:
import numpy as np

mat1 = np.array([[1, 2], [3, 4]])
mat2 = np.array([[5, 6], [7, 8]])

# Penjumlahan
print(mat1 + mat2)

# Pengurangan
print(mat2 - mat1)

[[ 6  8]
 [10 12]]
[[4 4]
 [4 4]]


- Perkalian pada Matriks

Perkalian matriks dapat dilakukan sebagai:

**Perkalian Element-wise:** Mengalikan elemen yang bersesuaian di dua matriks (operator * di NumPy).

**Perkalian Matriks:** Perkalian linear standar di mana elemen dihitung sebagai hasil perkalian baris dan kolom (menggunakan fungsi np.dot() atau operator @).


In [ ]:
import numpy as np

mat1 = np.array([[1, 2], [3, 4]])
mat2 = np.array([[5, 6], [7, 8]])

# Perkalian element-wise
print(mat1 * mat2)

# Perkalian matriks
print(np.dot(mat1, mat2))

[[ 5 12]
 [21 32]]
[[19 22]
 [43 50]]


## **2.5 Indexing dan Slicing**

Seperti list Python, NumPy array mendukung akses elemen individu dan subset elemen menggunakan indeks. Fitur ini sangat fleksibel, bahkan memungkinkan manipulasi sebagian data tanpa mempengaruhi data aslinya.

In [ ]:
array = np.array([1, 2, 3, 4, 5])

# Mengakses elemen
print(array[0])
print(array[-1])

# Slicing
print(array[1:4])
print(array[:3])

1
5
[2 3 4]
[1 2 3]


## **2.6 Mengubah Bentuk (Reshape)**

Reshaping digunakan untuk mengubah bentuk array tanpa mengubah datanya. Misalnya, Anda dapat mengubah array 1D menjadi 2D atau sebaliknya.

Fungsi Utama:

- reshape(): Mengubah bentuk array menjadi dimensi yang diinginkan.
- flatten(): Mengubah array multi-dimensi menjadi 1D.


In [ ]:
import numpy as np

arr = np.array([1, 2, 3, 4, 5, 6])

# Mengubah menjadi array 2D (2 baris, 3 kolom)
reshaped_arr = arr.reshape(2, 3)
print("Reshaped Array:")
print(reshaped_arr)

# Mengubah kembali ke array 1D
flattened_arr = reshaped_arr.flatten()
print("\nFlattened Array:")
print(flattened_arr)

Reshaped Array:
[[1 2 3]
 [4 5 6]]

Flattened Array:
[1 2 3 4 5 6]


## **2.7 Menggabungkan Array (Concatenation)**

Penggabungan array berarti menggabungkan beberapa array menjadi satu, baik secara horizontal (dalam baris) maupun vertikal (dalam kolom).

Fungsi Utama:

- np.concatenate(): Menggabungkan array pada axis tertentu.
- np.vstack(): Menggabungkan secara vertikal (baris bertambah).
- np.hstack(): Menggabungkan secara horizontal (kolom bertambah)

In [ ]:
arr1 = np.array([[1, 2], [3, 4]])
arr2 = np.array([[5, 6], [7, 8]])

# Menggabungkan secara vertikal
vertical_concat = np.vstack((arr1, arr2))
print("Vertical Concatenation:")
print(vertical_concat)

# Menggabungkan secara horizontal
horizontal_concat = np.hstack((arr1, arr2))
print("\nHorizontal Concatenation:")
print(horizontal_concat)

Vertical Concatenation:
[[1 2]
 [3 4]
 [5 6]
 [7 8]]

Horizontal Concatenation:
[[1 2 5 6]
 [3 4 7 8]]


## **2.8 Pemisahan (Splitting)**


Pemisahan array berarti membagi satu array menjadi beberapa array kecil dengan jumlah elemen tertentu.

Fungsi Utama:

- np.split(): Membagi array pada index tertentu.
- np.vsplit(): Membagi secara vertikal.
- np.hsplit(): Membagi secara horizontal.

In [ ]:
arr = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])

# Membagi secara horizontal menjadi dua bagian
horizontal_split = np.hsplit(arr, 2)
print("Horizontal Splitting:")
print(horizontal_split[0])  # Bagian pertama
print(horizontal_split[1])  # Bagian kedua

# Membagi secara vertikal menjadi dua bagian
vertical_split = np.vsplit(arr, 2)
print("\nVertical Splitting:")
print(vertical_split[0])  # Bagian pertama
print(vertical_split[1])  # Bagian kedua

Horizontal Splitting:
[[1 2]
 [5 6]]
[[3 4]
 [7 8]]

Vertical Splitting:
[[1 2 3 4]]
[[5 6 7 8]]


# **3. Pandas Dataframe**

Pandas adalah library Python yang sangat berguna untuk analisis dan pengolahan data, cocok untuk menangani dataset dari ukuran menengah hingga besar. Dalam Pandas, terdapat dua jenis struktur data utama:

a. Series: Array satu dimensi yang dapat menyimpan berbagai jenis data dan memiliki label axis.

b. DataFrame: Struktur data dua dimensi, mirip dengan tabel dengan baris dan kolom, yang dapat menyimpan berbagai jenis data dan juga memiliki label axis.

Beberapa keunggulan dan fitur utama dari Pandas adalah:

a. Pemuatan Data yang Cepat: Dapat dengan mudah memuat data dari berbagai sumber seperti file CSV, Excel, atau database.

b. Manipulasi Data yang Efisien: Menawarkan berbagai fungsi dan metode untuk membersihkan, menyortir, dan menyusun ulang data dengan cara yang sangat efektif.

c. Penggabungan Data: Memungkinkan penggabungan dan pengelompokan data dengan mudah, serupa dengan SQL join operations.

d. Berbagai Fungsi Bawaan: Menyediakan banyak fungsi untuk mengagregasi, transformasi, dan analisis data yang komprehensif.


## **3.1  Membuat Dataframe dari Array Sederhana**


Membuat Dataframe dari Array Sederhana adalah salah satu cara dasar dalam menggunakan library Pandas untuk memanipulasi dan menganalisis data.

In [ ]:
import pandas as pd

# Memanggil dataframe constructor
dataframe = pd.DataFrame()
print(dataframe) # Dataframe kosong

# Membuat list berisi string
words = ['ini', 'adalah', 'dataframe']

# Memanggil dataframe constructor pada list
dataframe = pd.DataFrame(words)
print("\n", dataframe)

Empty DataFrame
Columns: []
Index: []

            0
0        ini
1     adalah
2  dataframe


## **3.2 Membuat Dataframe dari Dictionary Pandas**

Dictionary di Python adalah struktur data yang terdiri dari pasangan kunci dan nilai, di mana kunci dapat digunakan sebagai nama kolom di dalam DataFrame. Dengan menggunakan dictionary, pengguna dapat dengan mudah mengorganisasi data dan membuat Dataframe yang rapi, memungkinkan manipulasi dan analisis data yang lebih efisien.

In [ ]:
import pandas as pd

myvar = pd.DataFrame()
print(myvar) # Dataframe kosong

mydataset = {
  'Nama': ["Andi", "Budi", "Cinta", "Doni", "Elsa"],
  'Tinggi': [168, 170, 150, 154, 188]
}

# Memanggil dataframe constructor pada list
myvar = pd.DataFrame(mydataset)
print(myvar)

Empty DataFrame
Columns: []
Index: []
    Nama  Tinggi
0   Andi     168
1   Budi     170
2  Cinta     150
3   Doni     154
4   Elsa     188


# **CODELAB**

**CODELAB 1**

a. Buatlah 12 data tuple yang berisi masing-masing 4 data int, float, dan string dengan urutan acak.

b. Ubahlah data tuple tersebut menjadi list.

c. Tambahkan 2 data baru di index sesuai dengan NIM akhir kalian (jika NIM akhir adalah 0, maka index ke-10).

d. Hapus tipe data float pada list.

e. Urutkan elemen pada list secara descending.

In [1]:
import pandas as pd

produk_dict = {'Nama Produk': ['Produk A', 'Produk B', 'Produk C'],
            'ID Produk': [101, 102, 103]}
produk_series = pd.Series(produk_dict)
print("Series Produk:")
print(produk_series)

data_dict = {
    'Nama': ['Ratih', 'Dina', 'Lia', 'Ida', 'Doni'],
    'Usia': [22, 23, 24, 25, 26],
    'Kota': ['Jember', 'Malang', 'Jombang', 'Tulungagung', 'Surabaya'],
    'Jumlah Produk': [88, 92, 95, 85, 90]
}
df = pd.DataFrame(data_dict)
print("\nDataframe:")
print(df)


df['Negara'] = ['Indonesia', 'Indonesia', 'Indonesia', 'Indonesia', 'Indonesia']
print("\nDataframe setelah menambahkan kolom baru:")
print(df)


df_urut = df.sort_values(by='Nama')
print("\nDataframe yang diurutkan berdasarkan Nama:")
print(df_urut)

Series Produk:
Nama Produk    [Produk A, Produk B, Produk C]
ID Produk                     [101, 102, 103]
dtype: object

Dataframe:
    Nama  Usia         Kota  Jumlah Produk
0  Ratih    22       Jember             88
1   Dina    23       Malang             92
2    Lia    24      Jombang             95
3    Ida    25  Tulungagung             85
4   Doni    26     Surabaya             90

Dataframe setelah menambahkan kolom baru:
    Nama  Usia         Kota  Jumlah Produk     Negara
0  Ratih    22       Jember             88  Indonesia
1   Dina    23       Malang             92  Indonesia
2    Lia    24      Jombang             95  Indonesia
3    Ida    25  Tulungagung             85  Indonesia
4   Doni    26     Surabaya             90  Indonesia

Dataframe yang diurutkan berdasarkan Nama:
    Nama  Usia         Kota  Jumlah Produk     Negara
1   Dina    23       Malang             92  Indonesia
4   Doni    26     Surabaya             90  Indonesia
3    Ida    25  Tulungagung         

**CODELAB 2**

a. Buatlah array 2D berdasarkan 6 digit terakhir NIM kalian.

b. Lakukan operasi pertambahan, pengurangan dan perkalian pada matriks yang kalian buat.

c. Ubahlah matriks tersebut menjadi bentuk 1D.

d. Lakukan slicing pada matriks.

In [2]:
import numpy as np

matrix = np.array([[3, 1], [1, 1], [7, 6]])
print(matrix)

#penjumlahan
print(matrix + 1)
#pengurangan
print(matrix - 1)
#perkalian
print(matrix * 2)

#mengubah ke 1D
flattened_arr = matrix.flatten()
print(flattened_arr)


print(matrix[0])

#slicing
print(matrix[1:2])

[[3 1]
 [1 1]
 [7 6]]
[[4 2]
 [2 2]
 [8 7]]
[[2 0]
 [0 0]
 [6 5]]
[[ 6  2]
 [ 2  2]
 [14 12]]
[3 1 1 1 7 6]
[3 1]
[[1 1]]


**CODELAB 3**

a. Buatlah sebuah series yang berisi dictionary yang memuat nama produk dan ID produk.

b. Buatlah sebuah dataframe yang berisi dictionary yang memuat 4 kolom dan setiap kolom berisi 5 data. Tidak boleh ada kesamaan antar praktikan.

c. Tambahkan 1 kolom baru.

d. Urutkan dataframe, misalnya berdasarkan abjad jika kolom berupa Nama.


In [3]:
list = ["Rani", "Febri", "Lina", "Tina", 153, 160, 170, 180, 1.2, 2.2, 3.4, 3.5]
print(list)

list.insert(7, "Databaru")
list.insert(6, 111)
print(list)

list.remove(1.2)
list.remove("Rani")
print(list)

list.reverse()
print(list)

['Rani', 'Febri', 'Lina', 'Tina', 153, 160, 170, 180, 1.2, 2.2, 3.4, 3.5]
['Rani', 'Febri', 'Lina', 'Tina', 153, 160, 111, 170, 'Databaru', 180, 1.2, 2.2, 3.4, 3.5]
['Febri', 'Lina', 'Tina', 153, 160, 111, 170, 'Databaru', 180, 2.2, 3.4, 3.5]
[3.5, 3.4, 2.2, 180, 'Databaru', 170, 111, 160, 153, 'Tina', 'Lina', 'Febri']


# **TUGAS**

**TUGAS 1**

Buatlah sebuah fungsi yang menerima list sebagai parameter dan melakukan operasi berikut:
1. Tambahkan 10 elemen numerik (int dan float) secara acak.
2. Hapus elemen di indeks sesuai NIM (ganjil/genap).
3. Balik urutan elemen dalam list.
4. Hapus elemen dengan nilai terbesar dari list tersebut.
5. Urutkan elemen pada list secara descending.

In [4]:
import random

def hitung_jumlah_elemen(list):
    # Menghitung jumlah elemen dalam list
    return len(list)

def tambah_elemen_acak(list):
    # Menambahkan 10 elemen numerik acak
    for _ in range(10):
        elemen_acak = random.choice([random.randint(1, 100), random.uniform(1.0, 100.0)])
        list.append(elemen_acak)
    return list

def hapus_elemen_genap(list):
    # Menghapus elemen pada indeks genap
    for i in range(len(list)-1, -1, -1):
        if i % 2 == 0:
            list.pop(i)
    return list

def balik_urutan(list):
    # Membalik urutan elemen dalam list
    list.reverse()
    return list

def hapus_elemen_terbesar(list):
    # Menemukan nilai terbesar dalam list
    elemen_terbesar = max(list)
    # Menghapus elemen pertama yang ditemukan dengan nilai terbesar
    list.remove(elemen_terbesar)
    return list

def urutkan_descending(list):
    # Mengurutkan elemen dalam list secara descending
    list.sort(reverse=True)
    return list

my_list = [1, 2, 3, 4, 5]
print("List awal:", my_list)

# Menghitung jumlah elemen
print("Jumlah elemen:", hitung_jumlah_elemen(my_list))

# Menambahkan 10 elemen acak
my_list = tambah_elemen_acak(my_list)
print("List setelah ditambah 10 elemen acak:", my_list)

# Menghapus elemen pada indeks genap
my_list = hapus_elemen_genap(my_list)
print("List setelah elemen pada indeks genap dihapus:", my_list)

# Membalik urutan elemen
my_list = balik_urutan(my_list)
print("List setelah dibalik urutannya:", my_list)

# Menghapus elemen terbesar
my_list = hapus_elemen_terbesar(my_list)
print("List setelah elemen terbesar dihapus:", my_list)

# Mengurutkan elemen secara descending
my_list = urutkan_descending(my_list)
print("List setelah diurutkan secara descending:", my_list)


List awal: [1, 2, 3, 4, 5]
Jumlah elemen: 5
List setelah ditambah 10 elemen acak: [1, 2, 3, 4, 5, 96, 75, 14, 31, 22, 15, 30.62499722751379, 25, 34, 8.71800944888771]
List setelah elemen pada indeks genap dihapus: [2, 4, 96, 14, 22, 30.62499722751379, 34]
List setelah dibalik urutannya: [34, 30.62499722751379, 22, 14, 96, 4, 2]
List setelah elemen terbesar dihapus: [34, 30.62499722751379, 22, 14, 4, 2]
List setelah diurutkan secara descending: [34, 30.62499722751379, 22, 14, 4, 2]


**TUGAS 2**

Buatlah sebuah tuple yang berisi dictionary dengan 10 pasangan key value mata kuliah dan nilai. Dari 10 pasangan key value tersebut, buatlah minimal 3 diantaranya memiliki nilai kosong. Lakukan operasi berikut:
1. Hapus entri yang tidak lengkap atau memiliki nilai kosong.
2. Hitung jumlah entri yang tersisa setelah penghapusan.
3. Urutkan pasangan key value tersebut berdasarkan nilai secara ascending.
4. Carilah mata kuliah dengan nilai tertinggi.
5. Hitunglah rata-rata semua nilai mata kuliah.


In [5]:
# Membuat tuple yang berisi dictionary dengan 10 pasangan key-value (mata kuliah dan nilai)
nilai_mata_kuliah = (
    {'Mata Kuliah': 'Matematika', 'Nilai': 85},
    {'Mata Kuliah': 'Fisika', 'Nilai': 90},
    {'Mata Kuliah': 'Kimia', 'Nilai': None},  # Nilai kosong
    {'Mata Kuliah': 'Biologi', 'Nilai': 78},
    {'Mata Kuliah': 'Sejarah', 'Nilai': 88},
    {'Mata Kuliah': 'Ekonomi', 'Nilai': None},  # Nilai kosong
    {'Mata Kuliah': 'Bahasa Indonesia', 'Nilai': 92},
    {'Mata Kuliah': 'Bahasa Inggris', 'Nilai': None},  # Nilai kosong
    {'Mata Kuliah': 'Pemrograman', 'Nilai': 95},
    {'Mata Kuliah': 'Pendidikan Pancasila', 'Nilai': 80}
)

# Menampilkan entri seluruhnya
print("DAFTAR NILAI MATA KULIAH SELURUHNYA:")
for item in nilai_mata_kuliah:
    print(item)

#menghapus entri yang memiliki nilai kosong (None)
nilai_mata_kuliah_bersih = tuple(item for item in nilai_mata_kuliah if item['Nilai'] is not None)

print("\nDAFTAR NILAI MATA KULIAH (TANPA NILAI KOSONG):")
for item in nilai_mata_kuliah_bersih:
    print(item)

#jumlah entri tersisa
jumlah_entri_tersisa = len(nilai_mata_kuliah_bersih)
print(f"\nJumlah entri yang tersisa setelah penghapusan: {jumlah_entri_tersisa}")

#mengurutkan nilai secara ascending
nilai_mata_kuliah_terurut = tuple(sorted(nilai_mata_kuliah_bersih, key=lambda x: x['Nilai']))
print("\nTuple setelah diurutkan berdasarkan nilai (ascending):")
for item in nilai_mata_kuliah_terurut:
    print(item)

#nilai tertinggi
nilai_tertinggi = max(nilai_mata_kuliah_bersih, key=lambda x: x['Nilai'])
print("\nMata kuliah dengan nilai tertinggi:")
print(f"'{nilai_tertinggi['Mata Kuliah']}' dengan nilai {nilai_tertinggi['Nilai']}")

#menghitung rata-rata nilai
nilai_list = [item['Nilai'] for item in nilai_mata_kuliah_bersih]
rata_rata_nilai = sum(nilai_list) / len(nilai_list)
print(f"Rata-rata nilai mata kuliah (seluruhnya): {rata_rata_nilai:.2f}")


DAFTAR NILAI MATA KULIAH SELURUHNYA:
{'Mata Kuliah': 'Matematika', 'Nilai': 85}
{'Mata Kuliah': 'Fisika', 'Nilai': 90}
{'Mata Kuliah': 'Kimia', 'Nilai': None}
{'Mata Kuliah': 'Biologi', 'Nilai': 78}
{'Mata Kuliah': 'Sejarah', 'Nilai': 88}
{'Mata Kuliah': 'Ekonomi', 'Nilai': None}
{'Mata Kuliah': 'Bahasa Indonesia', 'Nilai': 92}
{'Mata Kuliah': 'Bahasa Inggris', 'Nilai': None}
{'Mata Kuliah': 'Pemrograman', 'Nilai': 95}
{'Mata Kuliah': 'Pendidikan Pancasila', 'Nilai': 80}

DAFTAR NILAI MATA KULIAH (TANPA NILAI KOSONG):
{'Mata Kuliah': 'Matematika', 'Nilai': 85}
{'Mata Kuliah': 'Fisika', 'Nilai': 90}
{'Mata Kuliah': 'Biologi', 'Nilai': 78}
{'Mata Kuliah': 'Sejarah', 'Nilai': 88}
{'Mata Kuliah': 'Bahasa Indonesia', 'Nilai': 92}
{'Mata Kuliah': 'Pemrograman', 'Nilai': 95}
{'Mata Kuliah': 'Pendidikan Pancasila', 'Nilai': 80}

Jumlah entri yang tersisa setelah penghapusan: 7

Tuple setelah diurutkan berdasarkan nilai (ascending):
{'Mata Kuliah': 'Biologi', 'Nilai': 78}
{'Mata Kuliah': 'Pendi

**TUGAS 3 (NIM GENAP)**

Buatlah dataframe yang terdiri dari 10 baris dan 6 kolom, dengan informasi sebagai berikut:

| | Brand | Cabang | Harga (ribu) | Stock | Status |
|--|---|---|---|---|---|
| 1 | Avoskin | Surabaya | 150 | 175 | Sold |
| 2 | G2G | Jakarta Pusat | 75 | 160 | Available |
| 3 | Skintific | Bandung | 120 | 120 | Available
| 4 | Wardah | Bandung | 85 | 300 | Sold |
| 5 | Emina | Jakarta | 75 | 230 | Available |
| 6 | Somethinc | Bandung | 120 | 200 | Available |
| 7 | Azarine | Surabaya | 50 | 180 |Available |
| 8 | Lacoco | Bandung | 90 | 100 | Sold |
| 9 | Y.O.U | Jakarta Pusat | 55 | 180 |Available |
| 10 | Madame Gie | Surabaya | 50 | 200 | Sold |



In [6]:
import pandas as pd

data = {
    'Brand': ['Avoskin', 'G2G', 'Skintific', 'Wardah', 'Emina', 'Somethinc', 'Azarine', 'Lacoco', 'Y.O.U', 'Madame Gie'],
    'Cabang': ['Surabaya', 'Jakarta Pusat', 'Bandung', 'Bandung', 'Jakarta', 'Bandung', 'Surabaya', 'Bandung', 'Jakarta Pusat', 'Surabaya'],
    'Harga (ribu)': [150, 75, 120, 85, 75, 120, 50, 90, 55, 50],
    'Stock': [175, 160, 120, 300, 230, 200, 180, 100, 180, 200],
    'Status': ['Sold', 'Available', 'Available', 'Sold', 'Available', 'Available', 'Available', 'Sold', 'Available', 'Sold']
}

# Membuat DataFrame
df = pd.DataFrame(data)

# Mengurutkan harga terlebih dahulu
sorted_prices = sorted(df['Harga (ribu)'])

# Menghitung kuartil setelah data diurutkan secara manual
q1_value = sorted_prices[len(sorted_prices) // 4]  # Q1 di posisi ke-25%
q2_value = sorted_prices[len(sorted_prices) // 2]  # Q2 di posisi ke-50%
q3_value = sorted_prices[(len(sorted_prices) * 3) // 4]  # Q3 di posisi ke-75%

# Menambahkan kolom kategori berdasarkan harga
def kategori_harga(row):
    if row['Harga (ribu)'] < q1_value:
        return 'Murah'
    elif row['Harga (ribu)'] > q3_value:
        return 'Mahal'
    else:
        return 'Sedang'

# Menambahkan kolom 'Kategori Harga'
df['Kategori Harga'] = df.apply(kategori_harga, axis=1)

# Menampilkan DataFrame dengan kolom 'Kategori Harga'
print(df[['Brand', 'Harga (ribu)', 'Kategori Harga']])

# Menampilkan statistik lainnya seperti nilai minimum, maksimum, jumlah stok, rata-rata, dll
min_values = df[['Harga (ribu)']].min()
print("[HARGA TERENDAH]", min_values.to_string())

max_values = df[['Harga (ribu)']].max()
print("[HARGA TERTINGGI]", max_values.to_string())

total_stock = df['Stock'].sum()
print("Jumlah total stok seluruh barang/produk:", total_stock)

mean_values = df[['Harga (ribu)', 'Stock']].mean()
print("Rata-rata Harga (ribu) dan Stock:")
print(mean_values.to_string())

print("Urutan Harga (terendah hingga tertinggi):")
print(sorted_prices)

# Q1, Q2, Q3
print("\nKuartil Pertama (Q1):", q1_value)
print("Kuartil Kedua (Q2 / Median):", q2_value)
print("Kuartil Ketiga (Q3):", q3_value)

df['Penghasilan'] = df['Harga (ribu)'] * df['Stock']
print("Seluruh Brand (Harga x Stock):")
print(df[['Brand', 'Harga (ribu)', 'Stock', 'Penghasilan']])

top_3_brands = df[['Brand', 'Penghasilan']].sort_values(by='Penghasilan', ascending=False).head(3)
print("\n3 Brand dengan Penghasilan Terbesar:")
print(top_3_brands)

grouped = df.groupby('Harga (ribu)')['Brand'].apply(list).reset_index()
print("Brand yang Dikelompokkan Berdasarkan Harga yang Sama:")
print(grouped)


        Brand  Harga (ribu) Kategori Harga
0     Avoskin           150          Mahal
1         G2G            75         Sedang
2   Skintific           120         Sedang
3      Wardah            85         Sedang
4       Emina            75         Sedang
5   Somethinc           120         Sedang
6     Azarine            50          Murah
7      Lacoco            90         Sedang
8       Y.O.U            55         Sedang
9  Madame Gie            50          Murah
[HARGA TERENDAH] Harga (ribu)    50
[HARGA TERTINGGI] Harga (ribu)    150
Jumlah total stok seluruh barang/produk: 1845
Rata-rata Harga (ribu) dan Stock:
Harga (ribu)     87.0
Stock           184.5
Urutan Harga (terendah hingga tertinggi):
[50, 50, 55, 75, 75, 85, 90, 120, 120, 150]

Kuartil Pertama (Q1): 55
Kuartil Kedua (Q2 / Median): 85
Kuartil Ketiga (Q3): 120
Seluruh Brand (Harga x Stock):
        Brand  Harga (ribu)  Stock  Penghasilan
0     Avoskin           150    175        26250
1         G2G            75    16

TypeError: unhashable type: 'list'

Lakukan analisis berikut pada DataFrame tersebut:
1. Tampilkan ringkasan statistik deskriptif yang mencakup nilai minimum, maksimum, jumlah, rata-rata, standar deviasi, serta nilai kuartil 1, 2, dan 3.
2. Tentukan 3 brand dengan penghasilan terbesar (harga x stock).
3. Kelompokkan brand dengan harga yang sama.
4. Buatlah kolom baru yang berisi kategori ‘Murah’, ‘Sedang’, dan ‘Mahal’. Kategori ini didasarkan pada aturan:
- Harga di bawah kuartil 1 dikategorikan sebagai ‘Murah’
- Harga di atas kuartil 3 dikategorikan sebagai ‘Mahal’
- Sisanya dikategorikan sebagai ‘Sedang’.


**TUGAS 3 (NIM GANJIL)**

Buatlah dataframe yang terdiri dari 10 baris dan 6 kolom, dengan informasi sebagai berikut:

| | Brand | Tahun Rilis | Terjual | Harga (juta) | Jenis Penjualan |
|--|---|---|---|---|---|
| 1 | Apple | 2015 | 150 | 19 | Online |
| 2 | Xiaomi | 2017 | 75 | 9 | Offline |
| 3 | Samsung | 2018 | 120 | 10 | Online |
| 4 | Infinix | 2015 | 85 | 9 | Online |
| 5 | Oppo | 2017 | 75 | 4 | Offline |
| 6 | Asus | 2020 | 120 | 9 | Online |
| 7 | Lenovo | 2019 | 50 | 15 |Offline |
| 8 | Huawei | 2019 | 90 | 15 | Online |
| 9 | Google  Pixel |  2020 | 55 | 15 |Offline |
| 10 | Sony | 2017 | 50 | 11 | Offline |



Lakukan analisis berikut pada DataFrame tersebut:
1. Tampilkan ringkasan statistik deskriptif yang mencakup nilai minimum, maksimum, jumlah, rata-rata, standar deviasi, serta nilai kuartil 1, 2, dan 3.
2. Tentukan 3 brand dengan penghasilan terbesar (harga x terjual).
3. Kelompokkan brand dengan jenis penjualan yang sama.
4. Buatlah kolom baru yang berisi kategori ‘Murah’, ‘Sedang’, dan ‘Mahal’. Kategori ini didasarkan pada aturan:
- Harga di bawah kuartil 1 dikategorikan sebagai ‘Murah’
- Harga di atas kuartil 3 dikategorikan sebagai ‘Mahal’
- Sisanya dikategorikan sebagai ‘Sedang’.


# **KRITERIA & DETAIL PENILAIAN**

| Detail | Bobot Nilai |
|---|---|
| Codelab | 15% |
| Tugas | 45% |
| Pemahaman Materi & Ketepatan Jawaban | 40% |